MG-1887 - Add support for OAuth2.0

rodneyosodo opened the pull request:
<!-- Copyright (c) Abstract Machines
SPDX-License-Identifier: Apache-2.0 -->

<!--

Pull request title should be `MG-XXX - description` or `NOISSUE - description` where XXX is ID of the issue that this PR relate to.
Please review the [CONTRIBUTING.md](https://github.com/absmach/magistrala/blob/master/CONTRIBUTING.md) file for detailed contributing guidelines.

For Work In Progress Pull Requests, please use the Draft PR feature, see https://github.blog/2019-02-14-introducing-draft-pull-requests/ for further details.

For a timely review/response, please avoid force-pushing additional commits if your PR already received reviews or comments.

- Provide tests for your changes.
- Use descriptive commit messages.
- Comment your code where appropriate.
- Squash your commits
- Update any related documentation.
-->

# What type of PR is this?

This is a feature PR because it adds Google OAuth2.0 as a provider. This has integrated OAuth2 authentication, specifically adding Google OAuth2 support. Changes include updating workflow configurations, adding OAuth-related fields to protocol buffers, and enhancing various components (GRPC, HTTP, JWT, services, and tests) to handle OAuth tokens and callbacks. New packages for OAuth2 functionality and Google integration, along with Docker and Nginx configurations for OAuth callbacks, have been introduced.

<!--This represents the type of PR you are submitting.

For example:
This is a bug fix because it fixes the following issue: #1234
This is a feature because it adds the following functionality: ...
This is a refactor because it changes the following functionality: ...
This is a documentation update because it updates the following documentation: ...
This is a dependency update because it updates the following dependencies: ...
This is an optimization because it improves the following functionality: ...
-->

## What does this do?

- **New Features**
    - Introduced OAuth2 integration for authentication, including support for Google OAuth2.
    - Added OAuth-related fields and functionality across various components to enhance security and user authentication.
- **Refactor**
    - Updated and refactored authentication services to support OAuth2 functionality.
- **Tests**
    - Added and updated tests to cover OAuth2 authentication and related functionalities.
- **Chores**
    - Updated dependencies and configurations to support OAuth2 integration.
- **Documentation**
    - Included OAuth2 setup instructions in the environment configuration files.
   
<!--
Please provide a brief description of what this PR is intended to do.
Include List any changes that modify/break current functionality.
-->

## Which issue(s) does this PR fix/relate to?

<!--
For pull requests that relate or close an issue, please include them below.  We like to follow [Github's guidance on linking issues to pull requests](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

For example having the text: "Resolves #1234" would connect the current pull request to issue 1234.  And when we merge the pull request, Github will automatically close the issue.
-->

Closes #1887.

## Have you included tests for your changes?

Yes, I have included tests for my changes.

<!--If you have not included tests, please explain why.
For example:
No, I have not included tests because I do not know how to.
-->

## Did you document any new/modified feature?

Yes, I have updated the documentation for the new feature.

<!--If you have not included documentation, please explain why.
For example:
No, I have not updated the documentation because I do not know how to.
-->

### Notes

<!--Please provide any additional information you feel is important.-->
Linked to https://github.com/absmach/magistrala-ui/pull/198

rodneyosodo committed:
add google as an oauth2.0 provider

Signed-off-by: Rodney Osodo <28790446+rodneyosodo@users.noreply.github.com>

rodneyosodo committed:
add tests

Signed-off-by: Rodney Osodo <28790446+rodneyosodo@users.noreply.github.com>

dborovcanin commented on auth/jwt/tokenizer.go:
```diff
@@ -32,25 +33,33 @@ var (
 
 type tokenizer struct {
 	secret []byte
+	google oauth2.Provider
```
Rename to something more generic; we will support other providers, as well.

dborovcanin commented on auth/jwt/tokenizer.go:
```diff
@@ -32,25 +33,33 @@ var (
 
 // NewRepository instantiates an implementation of Token repository.
-func New(secret []byte) auth.Tokenizer {
+func New(secret []byte, google oauth2.Provider) auth.Tokenizer {
```
Same here.

dborovcanin commented on auth/jwt/tokenizer.go:
```diff
@@ -64,6 +73,13 @@ func (repo *tokenizer) Issue(key auth.Key) (string, error) {
 		Expiration(key.ExpiresAt)
 	builder.Claim(userField, key.User)
 	builder.Claim(domainField, key.Domain)
+	if key.OAuth.Provider == repo.google.Name() {
```
Not related to this PR, but please rename the pointer receiver to `t` instead of `repo`. It's confusing.

dborovcanin commented on auth/jwt/tokenizer.go:
```diff
@@ -125,5 +141,63 @@ func (repo *tokenizer) Parse(token string) (auth.Key, error) {
+			}
+
+			return key, nil
+		} else {
```
No need for this `else`.

rodneyosodo committed:
Refactored test files to use non nil oauth2mock.

Updated setup functions to pass oauth2mocks.Provider as parameter to api.MakeHandler. Added HTTP server setup, mock service initialization, logger setup, user API endpoints handler, and OAuth2 provider for authentication.

Signed-off-by: Rodney Osodo <28790446+rodneyosodo@users.noreply.github.com>

rodneyosodo committed:
enhance godocs and add default timeout for http clients

Handle OAuth2.0 errors returned during registration and signing in
process
Signed-off-by: Rodney Osodo <28790446+rodneyosodo@users.noreply.github.com>

dborovcanin commented on cmd/users/main.go:
```diff
@@ -172,8 +177,16 @@ func main() {
 		return
 	}
 
+	googleConfig := oauth2.Config{}
```
Use generic names, so we can easily replace the implementation later.

rodneyosodo replied:
This is google specific config because we have prefixed it with `MG_GOOGLE_`

dborovcanin replied:
This is Oauth2 config, it is not necessarily google-specific. The only specifc thing should be the implementation we have in package with `googleoauth` alias.

rodneyosodo replied:
For example, if we add GitHub we will have 2 config files with different prefixes because they have different `clientIDs` and `clientSecrets`

dborovcanin replied:
If we go with multiple providers, we will use a slice/array anyway.

dborovcanin commented on cmd/users/main.go:
```diff
@@ -50,29 +52,32 @@ import (
+	ESURL              string  `env:"MG_ES_URL"                       envDefault:"nats://localhost:4222"`
+	TraceRatio         float64 `env:"MG_JAEGER_TRACE_RATIO"           envDefault:"1.0"`
+	SelfRegister       bool    `env:"MG_USERS_ALLOW_SELF_REGISTER"    envDefault:"false"`
+	OAuthUIRedirectURL string  `env:"MG_OAUTH_UI_REDIRECT_URL"        envDefault:"http://localhost:9095/domains"`
+	OAuthUIErrorURL    string  `env:"MG_OAUTH_UI_ERROR_URL"           envDefault:"http://localhost:9095/error"`
```
No need for `UI` in env var name.

rodneyosodo replied:
These are UI-related URLs which is why I added UI to them. It is not similar to the OAuth2.0 generic redirect URL. 
- `MG_OAUTH_UI_REDIRECT_URL` is used to redirect the user to UI after successful authentication
- `MG_OAUTH_UI_ERROR_URL` is used to redirect the user to UI when we encounter any error in the OAuth2.0 flow

dborovcanin commented on pkg/oauth2/google/oauth2.go:
```diff
@@ -0,0 +1,181 @@
+}
+
+func (cfg *config) Refresh(ctx context.Context, token string) (oauth2.Token, error) {
+	payload := strings.NewReader("grant_type=refresh_token&refresh_token=" + token + "&client_id=" + cfg.config.ClientID + "&client_secret=" + cfg.config.ClientSecret)
```
Use `Sprintf` or string builder.

dborovcanin commented on .github/workflows/check-generated-files.yml:
```diff
@@ -62,7 +62,7 @@ jobs:
         if: steps.changes.outputs.proto == 'true'
         run: |
-          PROTOC_VERSION=25.1
+          PROTOC_VERSION=25.2
```
Update `PROTOC_GEN_VERSION` to `v1.32.0`.

rodneyosodo replied:
The latest version is `25.3` https://github.com/protocolbuffers/protobuf/releases/tag/v25.3

dborovcanin replied:
I am talking about `protoc-gen` plugin for Go: [https://github.com/protocolbuffers/protobuf-go](https://github.com/protocolbuffers/protobuf-go/releases)

rodneyosodo committed:
make factory function more generic

Signed-off-by: Rodney Osodo <28790446+rodneyosodo@users.noreply.github.com>

dborovcanin commented on users/api/clients.go:
```diff
@@ -515,3 +521,64 @@ func queryPageParams(r *http.Request, defPermission string) (mgclients.Page, err
+		var state string
+		var flow oauth2.State
+		var err error
+		if r.FormValue("state") != "" && strings.Contains(r.FormValue("state"), "-") {
```
The first check is not needed.

rodneyosodo committed:
Fix tests for token with OAuth

- Implement tokenizer struct with secret key and OAuth provider map
- Handle OAuth providers and tokens, and perform error checking

Signed-off-by: Rodney Osodo <28790446+rodneyosodo@users.noreply.github.com>

rodneyosodo committed:
Updated version numbers for dependencies

Updated protoc_gen_go version and oauthConfig.

Signed-off-by: Rodney Osodo <28790446+rodneyosodo@users.noreply.github.com>

dborovcanin commented on auth/jwt/tokenizer.go:
```diff
@@ -125,5 +151,66 @@ func (repo *tokenizer) Parse(token string) (auth.Key, error) {
+	if ok {
+		claims, ok := oauthToken.(map[string]interface{})
+		if !ok {
+			return auth.Key{}, errors.Wrap(fmt.Errorf("invalid claims for %s token", provider.Name()), ErrParseToken)
```
Reverse wrapping order - wrap with known error (ErrParseToken).

dborovcanin replied:
Do the same for other cases.

rodneyosodo committed:
fix(auth/jwt): fix error handling in parseOAuthToken

Refactored error handling in the parseOAuthToken function to correctly wrap the errors for invalid claims related to access and refresh tokens. This change ensures proper error messaging and handling in the JWT token parsing process.

Signed-off-by: Rodney Osodo <28790446+rodneyosodo@users.noreply.github.com>

rodneyosodo committed:
rename files from oauth2.go to provider.go

Signed-off-by: Rodney Osodo <28790446+rodneyosodo@users.noreply.github.com>

dborovcanin commented on pkg/oauth2/oauth2.go:
```diff
@@ -0,0 +1,76 @@
+	// IsEnabled checks if the OAuth2 provider is enabled.
+	IsEnabled() bool
+	// Profile retrieves the user's profile and OAuth tokens from the OAuth2 provider.
+	Profile(ctx context.Context, code string) (mfclients.Client, oauth2.Token, error)
```
Rename to `UserDetails`.

rodneyosodo committed:
rename profile to userdetails

Signed-off-by: Rodney Osodo <28790446+rodneyosodo@users.noreply.github.com>

dborovcanin commented on pkg/oauth2/google/provider.go:
```diff
@@ -0,0 +1,182 @@
+}
+
+func (cfg *config) Name() string {
+	return "google"
```
Let's extract to constant.

dborovcanin commented on pkg/oauth2/google/provider.go:
```diff
@@ -0,0 +1,182 @@
+		},
+		Metadata: map[string]interface{}{
+			"oauth_provider":  "google",
+			"profile_picture": user.Picture,
```
Use constants.

rodneyosodo committed:
add providerName as a constant

Signed-off-by: Rodney Osodo <28790446+rodneyosodo@users.noreply.github.com>

dborovcanin commented on pkg/oauth2/oauth2.go:
```diff
@@ -0,0 +1,76 @@
+type Provider interface {
+	// Name returns the name of the OAuth2 provider.
+	Name() string
+	// State returns the current state for the OAuth2 flow.
```
Add empty lines below each method (align with other interfaces).
